csrc/executor.cpp

```diff
@@ -982,21 +986,51 @@ std::vector<at::Tensor> allocateOutputs(
     ExpressionEvaluator& ee) {
   FUSER_PERF_SCOPE("allocateOutputs");
 
-  std::vector<at::Tensor> outputs;
-  outputs.reserve(output_info.size());
-
-  for (const auto output_idx : c10::irange(output_info.size())) {
-    Val* out = kernel->outputs()[output_idx];
-    auto [aliased_in, alias_info] = kernel->getOutputAlias(out);
-    auto out_tensor = allocateOutput(
-        output_info[output_idx], aliased_in, alias_info, device, ee);
-    // Bind `out_tensor` so duplicated outputs map to the same tensor.
+  const auto num_outs = output_info.size();
+
+  // Sort the outputs so we compute aliases after allocating non-aliases. The
+  // order between aliases can be arbitrary. E.g.,
+  //
+  // ```
+  // non_alias_out = ...
+  // alias_out_0 = reshape(non_alias_out, ...)
+  // alias_out_1 = reshape(alias_out_0, ...)
+  // ```
+  //
+  // It's fine to compute `alias_out_1` before computing `alias_out_0`: when we
+  // compute `alias_out_1`, `alias_out_0` will be recursively
+  // `ExpressionEvaluator::evaluate`ed. However, `non_alias_out` must be
+  // allocated first.
+  std::vector<std::pair<int64_t, Val*>> sorted_outs;
```

Comment:
Just want to make sure I understand this correctly, this sorting is not sorting the outputs themselves but just the order of their allocations, right?

Reply:
Yes!

Reply:
nitpick: maybe a note there for us mortals that don't get this quickly enough, that non-aliased outputs needs to be allocated since allocation of aliased tensor needs to reference them.